### PR TITLE
Restructured code of againstElectronDeadECAL tauID (10_2_X)

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:
Restructured code of againstElectronDeadECAL discriminator to run on either AOD or miniAOD inputs (backport of #29627):
- add a version of the discriminant to run on top of pat::Taus to be used with miniAOD data samples;
- use eta and phi of impact position of the leading PFChargedHadron on ECal surface instead of eta and phi of tau at the vertex (taking bending of track in the magnetic field and the zVertex into account) - feature **disabled by default** for backwards compatibility; 
- restructure the code in the way that both AOD and miniAOD version of the discriminant call the same algorithm to avoid code duplication;
- move code performing the extrapolation to the ECal surface from the againstElectronMVA algorithm to an auxiliary helper class which avoids code duplication;

Expected some differences:
- none with default workflows

#### PR validation:

`runTheMatrix.py -l limited -i all` passed successfully except a few failed due to problem with inputs (DAS_ERROR).

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

It is backport of #29627

### CODE REVIEW

**Comment by perrotta** [comment]
backport of #29627

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstElectronDeadECAL.cc [hunk 1]
```diff
@@ -0,0 +1,87 @@
+/** \class TauDiscriminationAgainstElectronDeadECALBase
```
Comment (slava77) [inline RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstElectronDeadECAL.cc:1]: this file needs an update

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc [hunk 2]
```diff
@@ -57,7 +57,6 @@ AntiElectronIDMVA6::AntiElectronIDMVA6(const edm::ParameterSet& cfg)
   Var_woGwGSF_Endcap_ = new Float_t[23];
   Var_wGwGSF_Endcap_ = new Float_t[31];
```
Comment (slava77) [inline RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc:58]: compared to the master/106X versions, this file is missing removal of 
```c++
#include "MagneticField/Engine/interface/MagneticField.h"
#include "MagneticField/Records/interface/IdealMagneticFieldRecord.h"
```

**Review by slava77 — COMMENTED** [review]
To be updated, to match the 106X or the master:
- RecoTauTag/RecoTau/interface/TauDiscriminationProducerBase.h getProducerString and similarly in TauDiscriminationProducerBase.cc
- RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstElectronDeadECAL.cc
- RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc

File: RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc (hunk 2 above)
Comment (mbluj) [reply RecoTauTag/RecoTau/src/AntiElectronIDMVA6.cc:58]: Also indentation is different => synchronized.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by santocch** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/29747	c7466f5 
- this backport PR is consistent with the master version; more close to the 106X version of the backport as in #29739 (except for no extra change added in PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py: the default behavior is unchanged and apparently there is no change to add the PAT/miniAOD-based  deadECAL discriminant in the run2_miniAOD_devel here
- jenkins tests pass and comparisons with the baseline show no  (relevant) differences

**Comment by silviodonato** [comment]
+1

**Pull request merged**